### PR TITLE
Tests pipeline functions

### LINKED ISSUE #396: [TESTS] - PipelineFunctions
Build unit tests for all functions in PipelineFunctions.scala

### PR DESCRIPTION
**Pull request by sawankulkarni-db**
closes #396 
Tests for functions in PipelineFunctions.scala
Please review and let me know for any improvements.
Thank you!

### CODE REVIEW

**Comment by GeekSheikh** [comment]
@sundarshankar89 @Sriram-databricks -- I have asked @sawankulkarni-db to add you guys as reviewers for this PR. I'd like for you to review the PR for test patterns and conformity. If you happen to see any scenario coverage issues, please bring them up but that is not the focus of your review. I will do a review to look for any pertinent missing scenarios but would like to get the team reviewing one-another's code.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala [hunk 1]
```diff
@@ -185,4 +191,199 @@ class PipelineFunctionsTest extends AnyFunSpec with DataFrameComparer with Spark
...
+    it("should throw an exception - empty string") {
+      Given("an event hub connection string")
+      val ehConnString = ""
+
+      When("the connection string is not in correct format")
+
+      Then("the function throws an exception")
+      assertThrows[BadConfigException] (PipelineFunctions.parseEHConnectionString(ehConnString))
+    }
+    it("should throw an exception - incorrect format") {
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala:208]: sometimes the SharedAccessKey has a + and sometimes it doesn't. We've had cases where the + broke our parser, can we add a test case for that to ensure we don't hit that again? Additionally, the sharedAccessKey format is published, perhaps we can test the edge cases of that format to help guard against future breaks. This is a very nasty and public facing break if it happens.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala [hunk 2]
```diff
@@ -185,4 +191,199 @@ class PipelineFunctionsTest extends AnyFunSpec with DataFrameComparer with Spark
...
+      val df = spark.sql("select '1636663343887' as epoch_millisecond")
+
+      When("function is called on this column")
+
+      Then("returns a column of type timestamp and preserves milliseconds")
+      assertResult("TimestampType") (df
+        .withColumn("converted_string", PipelineFunctions.epochMilliToTs("epoch_millisecond"))
+        .dtypes.filter(_._1 == "converted_string").head._2
+      )
+      assertResult("887") (df
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala:254]: why 887? Sorry I'm not getting it. Can we be a little more verbose in the when / then and/or add some comments?

File: src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala [hunk 3]
```diff
@@ -185,4 +191,199 @@ class PipelineFunctionsTest extends AnyFunSpec with DataFrameComparer with Spark
...
+        .withColumn("converted_column", PipelineFunctions.epochMilliToTs("epoch_millisecond"))
+        .select("converted_column")
+        .collect().head.get(0).toString.split("\\.").takeRight(1).head
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala:257]: this kind of logic is tough to follow. It's often necessary but without having an example of the input object format or comments users have to go digging into the code and reviewing input types to decipher what's going on. Let's try to make this more obvious.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala [hunk 4]
```diff
@@ -185,4 +191,199 @@ class PipelineFunctionsTest extends AnyFunSpec with DataFrameComparer with Spark
...
+      val testTable: PipelineTable = PipelineTable(name = "table1", _keys = Array("id"), config = dummyConfig)
+
+      When("function is called on the given delta table")
+      val functionOutputDf: DataFrame = PipelineFunctions.getDeltaHistory(spark, testTable)
+
+      Then("returns a dataframe with detal table history")
+      val expectedDf: DataFrame = DeltaTable.forPath(testTable.tableLocation).history(9999)
+        .select("version", "timestamp", "operation", "clusterId", "operationMetrics", "userMetadata")
+
+      assertApproximateDataFrameEquality(functionOutputDf, expectedDf, 0.001)
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala:326]: should we use `assertSmallDatasetEquality` instead of approximate? This is a tiny testDF I think approximate may not be appropriate here.

File: src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala [hunk 5]
```diff
@@ -185,4 +191,199 @@ class PipelineFunctionsTest extends AnyFunSpec with DataFrameComparer with Spark
...
+      When("function is called on the delta table that is never optimized")
+
+      Then("return 0")
+      assertResult(0L) (PipelineFunctions.getLastOptimized(spark, testTable))
+
+      // delete the delta table
+      val dir = Directory("/tmp/overwatch/tests/table1")
+      dir.deleteRecursively()
+    }
+    it("should throw an exception when path is not a delta table") {
```
Comment (GeekSheikh) [inline src/test/scala/com/databricks/labs/overwatch/pipeline/PipelineFunctionsTest.scala:369]: This is a duplicate test underneath a different function. Should we create some generic test suite for common, reusable tests like this?

**Comment by GeekSheikh** [comment]
@sawankulkarni-db -- you opened this against main, please be sure to open your PRs against the correct branch in this case 0612_release

**Comment by sundarshankar89** [comment]
Ready for review @GeekSheikh

**Review by GeekSheikh — APPROVED** [review]
Thank you @sundarshankar89 looks good. Please note that there were some changes to the parseEHConnectionString. @alexott We'll be merging this today, could you please take a look at these tests and either remove the ones that are no longer valid and/or update them if you don't already have sufficient tests in place. Thank you.

**Comment by sundarshankar89** [comment]
@alexott or @GeekSheikh tag me once this is merged, I will accommodate.
